### ISSUE TITLE
compiler crash: NeedsSourceLocation

### ISSUE THREAD

**Issue by LordMZTE** [ziglang/zig#16740]
### Zig Version

0.12.0-dev.21+ac95cfe44

### Steps to Reproduce and Observed Behavior

1. Create the file `bug.zig` with this content:
```zig
const std = @import("std");

test {
    const my_set = ComptimeStringSet(&.{ "a", "b" });
    _ = my_set;
}

pub fn ComptimeStringSet(comptime vals: []const []const u8) type {
    var kvs = [_]struct { []const u8, void }{};
    for (vals) |val|
        kvs = kvs ++ .{ val, {} };
    return std.ComptimeStringMap(void, kvs);
}
```

2. `zig test bug.zig`

3.
```
thread 13558 panic: zig compiler bug: NeededSourceLocation
Unable to dump stack trace: debug info stripped
fish: Job 1, 'zig test bug.zig' terminated by signal SIGABRT (Abort)
```

### Expected Behavior

The code compiles.

**Comment by ianprime0509** [comment]
This seems to be related to #16720.

<details>

```
thread 1126522 panic: zig compiler bug: NeededSourceLocation
/home/ian/src/zig/src/Sema.zig:2401:13: 0x5bb7305 in failWithOwnedErrorMsg (zig)
            return error.NeededSourceLocation;
            ^
/home/ian/src/zig/src/Sema.zig:27581:5: 0x5e16b79 in coerceExtra (zig)
    return sema.failWithOwnedErrorMsg(msg);
    ^
/home/ian/src/zig/src/Sema.zig:26857:21: 0x5bd2b63 in coerce (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Sema.zig:13367:47: 0x60a6062 in zirArrayCat (zig)
                const coerced_elem_val_inst = try sema.coerce(block, resolved_elem_ty, elem_val_inst, .unneeded);
                                              ^
/home/ian/src/zig/src/Sema.zig:987:46: 0x5decf36 in analyzeBodyInner (zig)
            .array_cat                    => try sema.zirArrayCat(block, inst),
                                             ^
/home/ian/src/zig/src/Sema.zig:902:21: 0x5bd9c17 in analyzeBodyBreak (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Sema.zig:1664:37: 0x5dfcccc in analyzeBodyInner (zig)
                const break_data = (try sema.analyzeBodyBreak(block, inline_body)) orelse
                                    ^
/home/ian/src/zig/src/Sema.zig:902:21: 0x5bd9c17 in analyzeBodyBreak (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Sema.zig:1557:37: 0x5dffa0d in analyzeBodyInner (zig)
                const break_data = (try sema.analyzeBodyBreak(block, inline_body)) orelse
                                    ^
/home/ian/src/zig/src/Sema.zig:902:21: 0x5bd9c17 in analyzeBodyBreak (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Sema.zig:1535:37: 0x5dfb81e in analyzeBodyInner (zig)
                const break_data = (try sema.analyzeBodyBreak(block, inline_body)) orelse
                                    ^
/home/ian/src/zig/src/Sema.zig:902:21: 0x5bd9c17 in analyzeBodyBreak (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Sema.zig:1557:37: 0x5dffa0d in analyzeBodyInner (zig)
                const break_data = (try sema.analyzeBodyBreak(block, inline_body)) orelse
                                    ^
/home/ian/src/zig/src/Sema.zig:885:21: 0x604a262 in analyzeBody (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Sema.zig:7159:33: 0x67c97c2 in analyzeCall (zig)
                    else => |e| return e,
                                ^
/home/ian/src/zig/src/Sema.zig:6642:9: 0x60b416f in zirCall__anon_141787 (zig)
        return sema.analyzeCall(block, func, func_ty, callee_src, call_src, modifier, ensure_result_used, resolved_args, bound_arg_src, call_dbg_node, .call);
        ^
/home/ian/src/zig/src/Sema.zig:1004:46: 0x5dedc93 in analyzeBodyInner (zig)
            .call                         => try sema.zirCall(block, inst, .direct),
                                             ^
/home/ian/src/zig/src/Sema.zig:5744:25: 0x6871c9a in resolveBlockBody (zig)
            else => |e| return e,
                        ^
/home/ian/src/zig/src/Sema.zig:5710:5: 0x618c897 in zirBlock (zig)
    return sema.resolveBlockBody(parent_block, src, &child_block, body, inst, &label.merges);
    ^
/home/ian/src/zig/src/Sema.zig:1545:32: 0x5dff743 in analyzeBodyInner (zig)
                    break :blk try sema.zirBlock(block, inst, tags[inst] == .block_comptime);
                               ^
/home/ian/src/zig/src/Sema.zig:885:21: 0x604a262 in analyzeBody (zig)
        else => |e| return e,
                    ^
/home/ian/src/zig/src/Module.zig:5328:39: 0x5dcd612 in analyzeFnBody (zig)
        error.NeededSourceLocation => @panic("zig compiler bug: NeededSourceLocation"),
                                      ^
/home/ian/src/zig/src/Module.zig:3922:40: 0x5ba11a6 in ensureFuncBodyAnalyzed (zig)
            var air = mod.analyzeFnBody(func_index, sema_arena) catch |err| switch (err) {
                                       ^
/home/ian/src/zig/src/Compilation.zig:3197:42: 0x5b9f260 in processOneJob (zig)
            module.ensureFuncBodyAnalyzed(func) catch |err| switch (err) {
                                         ^
/home/ian/src/zig/src/Compilation.zig:3134:30: 0x5a1569d in performAllTheWork (zig)
            try processOneJob(comp, work_item, main_progress_node);
                             ^
/home/ian/src/zig/src/Compilation.zig:2058:31: 0x5a11181 in update (zig)
    try comp.performAllTheWork(main_progress_node);
                              ^
/home/ian/src/zig/src/main.zig:4030:24: 0x5a3fc20 in updateModule (zig)
        try comp.update(main_progress_node);
                       ^
/home/ian/src/zig/src/main.zig:3451:17: 0x5a6322a in buildOutputType (zig)
    updateModule(comp) catch |err| switch (err) {
                ^
/home/ian/src/zig/src/main.zig:276:31: 0x58c7ab7 in mainArgs (zig)
        return buildOutputType(gpa, arena, args, .zig_test);
                              ^
/home/ian/src/zig/src/main.zig:214:20: 0x58c4bb5 in main (zig)
    return mainArgs(gpa, arena, args);
                   ^
/home/ian/src/zig/lib/std/start.zig:574:37: 0x58c459e in main (zig)
            const result = root.main() catch |err| {
                                    ^
???:?:?: 0x7f0e7879584f in ??? (libc.so.6)
Unwind information for `libc.so.6:0x7f0e7879584f` was not available, trace may be incomplete

Aborted (core dumped)
```

<summary>Debug compiler output on 0.12.0-dev.17+49244dc0c</summary>
</details>

**Comment by wooster0** [comment]
I can't test right now but I'm guessing this is fixed by #16517.
